### PR TITLE
feat: gradual rebalance, cluster compose example and scaling docs

### PR DESCRIPTION
**Pull request by gabrieljablonski**
## Summary

Stacked on #333 — closes out the implementation plan for #315.

### Gradual rebalance

A worker above the cluster fair share (`ceil(phones / liveInstances)` + tolerance) sheds **one connection per interval** (default 10s) to the least loaded peer. Designed to never destabilize:

- **Ordered handoff**: socket discarded → release cooldown → directed tombstone → lease released. The next owner can never overlap a still-open socket, so a migration produces zero `conflict/replaced` events.
- **Directed tombstone** names the intended next owner; the releaser's own claim loop (the lowest-latency claimant) skips it, and the target bypasses the fair-share cap (it was chosen because it is underloaded). Tombstone expiry falls back to an open claim, so a dead target never strands a phone.
- **Circuit breaker**: no rebalancing while recent claims indicate an in-progress failover, while draining, or while Redis is degraded.
- **Victim selection**: most recently claimed paired connection (least accumulated session warmth); pending-QR pairings are never moved.

This resolves the 1→2 migration scenario from the issue discussion: the older instance boots first and owns everything; once the new worker joins, connections equalize gradually with a single brief reconnect each (~8min for 100 connections at defaults, tunable).

### Deploy & docs

- `docker-compose.cluster.yml`: 1 proxy (only exposed port) + 2 workers (own media volumes) + Redis with `appendonly yes` / `everysec`; `stop_grace_period` > shutdown timeout; healthchecks on `/cluster/health`.
- README / README-pt "Scaling Horizontally": roles, behavior under worker crash / new worker / rolling deploy / Redis outage, operational requirements (AOF, worker reachability, pending-QR semantics), and the first-deploy caveat (the pre-lease container can't participate in the protocol, so that one rollout still shows legacy churn).

## Tests

Rebalance matrix in `coordinator.spec.ts`: safe ordering (discard/handoff before release), one-per-interval rate limit, tolerance band, no-peers, failover deferral, pending-QR exclusion; claim-cycle tombstone handling (skip when directed elsewhere, claim past the cap when directed at self).

`bun test`: 343 pass. `tsc --noEmit` and `biome check`: clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

- - -
This change is [Reviewable](https://reviewable.io/reviews/fazer-ai/baileys-api/334)

## Summary by CodeRabbit

* **New Features**
  * Optional cluster rebalancing: rate-limited, fair-share connection handoffs between instances with directed handoff support.

* **Documentation**
  * Multi-instance scaling guide added (English + Portuguese): roles, proxy+worker topology, handoff/failover, rolling deploys, Redis requirements, pairing caveats, operational notes.

* **Chores**
  * Cluster Docker Compose example and three rebalancing environment variables added.

* **Tests**
  * Added tests for rebalance, handoff, claim/release, and edge cases.

* **Bug Fixes**
  * Improved proxy route-cache subscription handling.

### CODE REVIEW

**Comment by gabrieljablonski** [comment]
@coderabbitai review

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: docker-compose.cluster.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      retries: 10
+
+  redis:
+    image: "redis:8-alpine"
+    command:
+      ["redis-server", "--appendonly", "yes", "--appendfsync", "everysec"]
+    volumes:
+      - "redis-data:/data"
+    restart: always
+
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply docker-compose.cluster.yml:124]: Fixed in 25713e8: the redis service now receives REDIS_PASSWORD and conditionally appends --requirepass via shell parameter expansion, so the example works both secured and unsecured. Verified the expansion with and without the variable set.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: src/cluster/coordinator.ts [hunk 2]
```diff
@@ -69,6 +72,15 @@ export class ClusterCoordinator {
   // lease. Drives the orphan override: the fair-share cap must never leave a
   // phone unowned forever (e.g. rounding when phones % instances != 0).
   private firstSeenUnleasedAt = new Map<string, number>();
+  // Claim recency per locally-held phone: rebalance releases the most
+  // recently claimed connection, minimizing churn on long-stable ones.
+  private claimedAt = new Map<string, number>();
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply src/cluster/coordinator.ts:77]: Fixed in 084c1b4: claimedAt.delete moved into releaseHeldLease, which every release path funnels through (drain bailout, reconnect failure, logout, shutdown handoff, connectWithLease failure), plus the renew-cycle self-fence path. The now-redundant delete in runRebalanceCycle was removed.

File: src/cluster/coordinator.ts [hunk 3]
```diff
@@ -333,6 +372,102 @@ export class ClusterCoordinator {
...
+  // and a circuit breaker that defers to in-progress failovers.
+  async runRebalanceCycle() {
+    if (!this.options.rebalanceEnabled || this.draining || this.redisDegraded) {
+      return;
+    }
+    const now = this.now();
+    if (
+      now - this.lastRebalanceReleaseAt <
+      this.options.rebalanceReleaseIntervalMs
+    ) {
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply src/cluster/coordinator.ts:391]: Fixed in 084c1b4: lastRebalanceReleaseAt now takes this.now() after the discard/release completes, so the throttle window starts when the release actually lands.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: docker-compose.cluster.yml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+#   that worker's phones).
+# - Redis MUST persist with AOF (appendonly): the Signal auth state lives
+#   there, and restoring a stale RDB snapshot regresses the crypto ratchet —
+#   which means re-pairing every number.
+# - Each worker keeps its own media volume; GET /media is routed by the
+#   proxy to the instance that holds the file.
+
+services:
+  proxy:
+    image: "ghcr.io/fazer-ai/baileys-api:latest"
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply docker-compose.cluster.yml]: Fixed in fc8c9d6: the three services now use ${BAILEYS_API_IMAGE_TAG:-latest}, a single variable for proxy and workers so a pinned deploy can never skew versions between them (a digest/hardcoded tag per service would allow exactly that). Default stays latest so the example remains copy-paste runnable, with a header note about pinning releases.

File: README-pt.md [hunk 5]
```diff
@@ -78,6 +78,32 @@ O `docker-compose.coolify.yml` pode ser adaptado para outros ambientes Docker. V
     - Armazene essas chaves com segurança e forneça-as no cabeçalho `x-api-key` para solicitações autenticadas.
     - Em desenvolvimento (`NODE_ENV=development`), a autenticação é ignorada.
 
+### Escalando Horizontalmente (Múltiplas Instâncias)
+
+Rodar mais de uma instância contra o mesmo Redis é suportado através de uma topologia proxy + workers. Cada identidade do WhatsApp é possuída via um **lease** no Redis (renovado periodicamente, com self-fencing em caso de perda), então duas instâncias nunca disputam o mesmo número com loops de `conflict/replaced`.
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply README-pt.md]: Fixed in fc8c9d6 with "Rodar mais de uma instância no mesmo Redis é suportado por meio de uma topologia proxy + workers." — keeps the original verb while fixing the agreement flagged by LanguageTool and replacing "através de".

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

**Comment by gabrieljablonski** [comment]
@coderabbitai resume

File: src/cluster/coordinator.ts [hunk 6]
```diff
@@ -333,6 +372,103 @@ export class ClusterCoordinator {
...
+    // owner can never overlap a still-open socket — a rebalance migration
+    // must never produce a 440.
+    await this.handler.discardConnection(victim);
+    await leaseStore.setReleaseCooldown(victim);
+    await leaseStore.setHandoffTarget(victim, target.instanceId);
+    // Compare-and-delete on the epoch held since the claim: if ownership
+    // already moved (or we re-acquired under a newer epoch), the release
+    // no-ops instead of dropping someone else's lease.
+    await this.releaseHeldLease(victim);
+    void registry.publishOwnershipChanged(victim);
```
Comment (gabrieljablonski) [inline src/cluster/coordinator.ts]: Fixed in 02ea14d: the cooldown and tombstone writes are wrapped in try/catch (they only steer where the phone lands next, so a failure degrades to an undirected release), and releaseHeldLease + publishOwnershipChanged now always run after a successful discard.

File: src/config.ts [hunk 7]
```diff
@@ -160,6 +163,18 @@ const config = {
       60_000,
       { min: 0 },
     ),
+    rebalanceEnabled: CLUSTER_REBALANCE_ENABLED !== "false",
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply src/config.ts]: Fixed in 02ea14d: added boolFromEnv (mirroring intFromEnv's fail-fast contract) and CLUSTER_REBALANCE_ENABLED now rejects anything other than "true"/"false" at boot. The README timing nitpick was applied in the same commit (both languages); the "rate limit" italics one was skipped — the rest of both READMEs uses English jargon unformatted, and a one-off italic would be inconsistent.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: src/cluster/coordinator.ts [hunk 8]
```diff
@@ -259,6 +296,8 @@ export class ClusterCoordinator {
         }
         this.heldLeaseEpochs.set(id, acquired.epoch);
         this.firstSeenUnleasedAt.delete(id);
+        this.claimedAt.set(id, now);
+        this.lastClaimAt = now;
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply src/cluster/coordinator.ts]: Fixed in 84d9fdb: claimedAt and lastClaimAt now take this.now() captured right after the successful acquireLease, so a long scan can no longer backdate them and let rebalance stop deferring mid-failover.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: src/cluster/coordinator.ts [hunk 9]
```diff
@@ -333,6 +375,114 @@ export class ClusterCoordinator {
...
+    // Failover storm in progress (we just claimed orphans): the cluster is
+    // not in a steady state, load numbers are moving — don't migrate on top.
+    if (
+      this.lastClaimAt > 0 &&
+      now - this.lastClaimAt < this.options.unclaimedGraceMs
+    ) {
+      return;
+    }
+
+    const instances = await registry.listLiveInstances().catch(() => []);
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply src/cluster/coordinator.ts]: Fixed in 43e62d0: the catch now warns with the same shape as runClaimCycle's message before falling back to []. The regression test nitpick was applied in the same commit (setHandoffTarget rejects, discard runs once, releaseLease still called once under the stored-lease epoch).

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: docker-compose.cluster.yml [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    environment:
+      - NODE_ENV=production
+      - ROLE=proxy
+      - REDIS_URL=redis://redis:6379
+      - REDIS_PASSWORD=${REDIS_PASSWORD}
+    depends_on:
+      redis:
+        condition: service_healthy
+    restart: always
+    healthcheck:
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply docker-compose.cluster.yml:41]: Fixed in 6e31114: stop_grace_period: 45s on the proxy, matching the workers.

File: src/cluster/coordinator.spec.ts [hunk 11]
```diff
@@ -315,6 +327,169 @@ describe("ClusterCoordinator", () => {
...
+      // an untracked claim).
+      expect(releaseLease).toHaveBeenCalledWith(expect.any(String), 7);
+      expect(setHandoffTarget).toHaveBeenCalledWith(
+        expect.any(String),
+        "peer-instance",
+      );
+      // Socket down → cooldown/tombstone → lease released. Never the reverse.
+      expect(order.indexOf("discard")).toBeLessThan(order.indexOf("release"));
+      expect(order.indexOf("handoff")).toBeLessThan(order.indexOf("release"));
+    });
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply src/cluster/coordinator.spec.ts:382]: Fixed in 6e31114: the safe-ordering test now records the cooldown write and asserts discard < cooldown < release alongside the existing handoff < release check.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: src/cluster/coordinator.ts [hunk 12]
```diff
@@ -333,6 +375,120 @@ export class ClusterCoordinator {
...
+        "[coordinator] rebalance handoff metadata failed for %s, releasing undirected: %s",
+        victim,
+        errorToString(error),
+      );
+    }
+    // Compare-and-delete on the epoch held since the claim: if ownership
+    // already moved (or we re-acquired under a newer epoch), the release
+    // no-ops instead of dropping someone else's lease.
+    await this.releaseHeldLease(victim);
+    void registry.publishOwnershipChanged(victim);
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply src/cluster/coordinator.ts]: Fixed in 814cafa: releaseHeldLease failure now sets redisDegraded (pausing further rebalances until a renewal clears it), warns, and returns; publishOwnershipChanged moved to a finally so the invalidation always fires. The blackhole window stays bounded by the lease TTL since the discarded victim left the renew set.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review
